stko/optimizers/collapser.py

```diff
@@ -0,0 +1,292 @@
+"""
+Collapser Optimizer
+===================
+
+#. :class:`.Collapser`
+
+Optimizer for collapsing enlarged topologies.
+
+"""
+
+import logging
+from itertools import combinations
+import numpy as np
+import uuid
+import os
+import shutil
+
+from .optimizers import Optimizer
+from ..utilities import get_atom_distance
+
+
+logger = logging.getLogger(__name__)
+
+
+class Collapser(Optimizer):
+    """
+    Collapse constructed molecules to decrease enlarged bonds.
+
+    This optimizer aims to bring extended bonds closer together for
+    further optimisation.
+
+    .. code-block:: python
+
+        import stk
+        import stko
+
+        bb1 = stk.BuildingBlock('NCCN', [stk.PrimaryAminoFactory()])
+        bb2 = stk.BuildingBlock(
+            smiles='O=CC(C=O)C=O',
+            functional_groups=[stk.AldehydeFactory()],
+        )
+        cage1 = stk.ConstructedMolecule(
+            topology_graph=stk.cage.FourPlusSix((bb1, bb2)),
+        )
+
+        # Perform collapser optimisation.
+        output_dir = f'cage_opt_{cage_name}_coll'
+        optimizer = stko.Collapser(
+            step_size=0.05,
+            distance_cut=2.0,
+            scale_steps=True,
+        )
+        cage1 = optimizer.optimize(mol=cage1)
+
+    """
+
+    def __init__(
+        self,
+        output_dir,
+        step_size,
+        distance_cut,
+        scale_steps=True,
+    ):
+        """
+        Initialize a :class:`MetalOptimizer` instance.
+
+        Parameters
+        ----------
+        output_dir : :class:`str`
+            The name of the directory into which files generated during
+            the calculation are written, if ``None`` then
+            :func:`uuid.uuid4` is used.
+
+        step_size : :class:`float`
+            The relative size of the step to take during collapse.
+
+        distance_cut : :class:`float`
+            Distance between distinct building blocks to use as
+            threshold for halting collapse in Angstrom.
+
+        scale_steps : :class:`bool`, optional
+            Whether to scale the step of each distict building block
+            by their relative distance from the molecules centroid.
+            Defaults to ``True``
+
+        """
+
+        self._output_dir = output_dir
+        self._step_size = step_size
+        self._distance_cut = distance_cut
+        self._scale_steps = scale_steps
+
+    def _get_inter_BB_distance(self, mol):
```

Comment:
I think this should have arg types. I have no idea whether this works for stk molecule or RDKit?

Reply:
Handle in the optimize function and class docs.